internal/cluster/store/fsm.go

```diff
@@ -0,0 +1,134 @@
+package store
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/hashicorp/raft"
+	"github.com/pbinitiative/zenbpm/internal/cluster/command/proto"
+	pb "google.golang.org/protobuf/proto"
+)
+
+// FSM is Finite State Machine of the system state
+type FSM struct {
+	s *Store
+}
+
+// NewFSM returns a new FSM.
+func NewFSM(s *Store) *FSM {
+	return &FSM{s: s}
+}
+
+var _ raft.FSM = &FSM{}
+
+// Apply applies a Raft log entry to the key-value store.
+func (f *FSM) Apply(l *raft.Log) interface{} {
+	f.s.mu.Lock()
```

Comment:
I understand the Mutex is meant to guard access to the `f.s.state`, yes?
If so, I believe it would be better to create the lock rather downwards in the call-tree, in the functions `applyNodeChange()` and `applyPartitionChange()` instead of upwards in the call-tree (here).
Reasoning:
* locks should be short living and local (general advice)
* in the function `applyNodeChange()` and `applyPartitionChange()` it's no more obvious the lock is there ... which is in contrast to the `Apply()` function has a lock()